clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -268,34 +283,41 @@ private RecordAppendResult tryAppend(long timestamp, byte[] key, byte[] value, H
     /**
      * Get a list of batches which have been sitting in the accumulator too long and need to be expired.
      */
-    public List<ProducerBatch> expiredBatches(int requestTimeout, long now) {
+    public List<ProducerBatch> expiredBatches(long deliveryTimeoutMs, long now) {
         List<ProducerBatch> expiredBatches = new ArrayList<>();
+
+        // Expiration of inflight batches happens here in the order of createdMs if the final state of the
+        // batch in not known by (batch's creation time + deliveryTimeoutMs).
+        while (!inFlightBatches.isEmpty() &&
+               now >= inFlightBatches.firstEntry().getKey() + deliveryTimeoutMs) {
+            Map.Entry<Long, List<ProducerBatch>> expiredInflightBatches = inFlightBatches.pollFirstEntry();
+            for (ProducerBatch batch: expiredInflightBatches.getValue()) {
+                if (batch.getFinalState() == null &&
+                    batch.maybeExpire(deliveryTimeoutMs, now)) {
```

Comment:
The check 'if (deliveryTimeoutMs <= (now - this.createdMs))' inside maybeExpire() would be true.
Looks like another method can be created inside ProducerBatch which expires the batch.

Reply:
`maybeExpire` has a side-effect of setting `errorMessage` internally. Hence calling it again in `if`.

Reply:
Understand.
That part can be refactored - goal is to reduce unnecessary comparison.

Reply:
@apurvam Those test don't even compile or run on my machine. What's up with those tests?

Reply:
They can't construct a kafka producer with the changes made in this PR.